cpp/src/arrow/compute/exec/asof_join_node.cc

```diff
@@ -1182,22 +1184,25 @@ class AsofJoinNode : public ExecNode {
 
       for (int i = 0; i < input_schema[j]->num_fields(); ++i) {
         const auto field = input_schema[j]->field(i);
+        bool as_output;
         if (i == on_field_ix) {
           ARROW_RETURN_NOT_OK(is_valid_on_field(field));
           // Only add on field from the left table
-          if (j == 0) {
-            fields.push_back(field);
-          }
+          as_output = (j == 0);
         } else if (std_has(by_field_ix, i)) {
           ARROW_RETURN_NOT_OK(is_valid_by_field(field));
           // Only add by field from the left table
-          if (j == 0) {
-            fields.push_back(field);
-          }
+          as_output = (j == 0);
         } else {
           ARROW_RETURN_NOT_OK(is_valid_data_field(field));
+          as_output = true;
+        }
+        if (as_output) {
           fields.push_back(field);
         }
+        if (field_output_indices) {
```

Comment:
What does this do?

Reply:
Imagine an asof join is called with:

LEFT: TIME, KEY1, KEY2, L1, L2
RIGHT: TIME, KEY1, KEY2, R1, R2

The output schema will be TIME, KEY1, KEY2, L1, L2, R1, R2

Substrait expects the output schema to be TIME, KEY1, KEY2, L1, L2, TIME, KEY1, KEY2, R1, R2

The field output indices will be 0, 1, 2, 3, 4, 0, 1, 2, 5, 6

The field output indices map from the schema substrait expects to the actual output schema.

This way, if the user asks for a field at index 5 (TIME) we can look up index 5 in the field output indices (which is 0) and return the item at index 0 in the output schema (which is TIME)